### PR TITLE
lk86/pipeline refactor

### PR DESCRIPTION
**Pull request by lk86**
Below is the list of modified files in this PR, separated by purpose. 

- Unify daemon, archive node, and rosetta builds, packages, and package publishing
  - Removed:
    - buildkite/src/Jobs/Release/ArchiveRedundancyTools.dhall
    - buildkite/src/Jobs/Release/ArchiveNodeArtifact.dhall
- Support Stretch and Buster packages, docker images, and CI jobs
  - New MinaArtifact jobs:
    - buildkite/src/Jobs/Release/MinaArtifactBuster.dhall
    - buildkite/src/Jobs/Release/MinaArtifactStretch.dhall
    - buildkite/src/Jobs/Release/MinaArtifact.dhall (deleted)
  - Two toolchain images:
    - buildkite/src/Jobs/Release/MinaToolchainArtifact.dhall
    - buildkite/src/Command/RunInToolchain.dhall
    - buildkite/src/Constants/ContainerImages.dhall
    - s/.minaToolchain/.minaToolchainBuster/ in a variety of dhall jobs
  - Dockerfiles updated to support buster or stretch dependencies
    - dockerfiles/Dockerfile-mina-archive
    - dockerfiles/Dockerfile-mina-daemon
  - Scripts updated to support determining dependencies from deb_codename
    - scripts/archive/build-release-archives.sh
    - scripts/rebuild-deb.sh
- Split Dockerfile-rosetta into individual stages, and turn Dockerfile-deb-builder and Dockerfile-toolchain into stages
  - Removed dockerfiles:
    - dockerfiles/Dockerfile-mina-daemon-puppeteered
    - dockerfiles/Dockerfile-rosetta
    - dockerfiles/Dockerfile-rosetta-hardcoded
    - dockerfiles/Dockerfile-toolchain
  - Re-target or remove the above dockerfiles from the list of containers in scripts/release-docker.sh:
    - scripts/release-docker.sh
  - New "Stages"
    - dockerfiles/stages/*
- Use buildkite/scripts/export-git-env-vars.sh everywhere instead of DOCKER_DEPLOY_ENV files
  - Removed logic:
    - buildkite/scripts/publish-deb.sh
  - New upload job that is used in many other downstream jobs
    - buildkite/src/Jobs/Release/GitEnvUpload.dhall
    - buildkite/scripts/export-git-env-vars.sh
    - Also many old variables are removed in favor of clearer names
- Make scripts/release-docker.sh and related dhall jobs more modular
  - Updated dhall job with parameters for deb_ codename, release, and version
    - buildkite/src/Command/DockerImage.dhall
  - Added support for these variables as explicit flags instead of `--extra-args --build_arg blah=blah2`
    - scripts/release-docker.sh
- Remove network-based archive node images as they work regardless of network ID
  - scripts/archive/build-release-archives.sh
- Remove opam.tar.gz caching and instead include all build-time dependencies in MinaToolchain
  -  dockerfiles/stages/2-toolchain
  - dockerfiles/stages/3-opam-deps
  - buildkite/src/Command/RunInToolchain.dhall
  - buildkite/src/Command/OpamInit.dhall (deleted)
  - Still some work remaining to automatically rebuild this when there are new dependencies. CI will build it, but you still have to manually update buildkite/src/Constants/ContainerImages.dhall

### CODE REVIEW

File: buildkite/scripts/generate-diff.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 # Base against origin/develop by default, but use pull-request base otherwise
```
Comment (bkase) [inline buildkite/scripts/generate-diff.sh:3]: can you update the comment to point to master by default too?

File: buildkite/src/Command/TestExecutive.dhall [hunk 2]
```diff
@@ -4,12 +4,12 @@ let Command = ./Base.dhall
 let Docker = ./Docker/Type.dhall
 let Size = ./Size.dhall
 
-let OpamInit = ../Command/OpamInit.dhall
```
Comment (bkase) [inline buildkite/src/Command/TestExecutive.dhall:7]: where does the opam init happen now? the docker container?

File: buildkite/src/Jobs/Release/MinaArtifactBuster.dhall [hunk 3]
```diff
@@ -0,0 +1,116 @@
+let Prelude = ../../External/Prelude.dhall
```
Comment (bkase) [inline buildkite/src/Jobs/Release/MinaArtifactBuster.dhall:1]: can we combine MinaArtifactBuster and MinaArtifactStretch into one function that can handle both?

I would recommend making a helper file and having a function that looks something like:

```dhall
let DebVersion = < Buster | Stretch >

let capitalName = \(debVersion : DebVersion) ->
  merge { Buster = "Buster", Stretch = "Stretch" } debVersion
 
let lowerName = \(debVersion : DebVersion) ->
  merge { Buster = "buster", Stretch = "stretch" } debVersion


let pipeline = \(debVersion : DebVersion) ->
  Pipeline.build
    ...
    name = "MinaArtifact${capitalName(debVersion)}"
    ...
    step_key="daemon-devnet-${lowerName(debVersion)}-docker-image"
    ...
    
let buster = pipeline Buster
let stretch = pipeline Stretch
```

and then you can use this in the MinaArtifactBuster.dhall files and MinaArtifactStretch.dhall files with:

```dhall
let ArtifactPipelines = ../../path/to/AritfactHelper.dhall in
ArtifactPipelines.buster
```

File: dockerfiles/stages/3-opam-deps [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#################################################################################################
+# The "opam-deps" Stage
```
Comment (bkase) [inline dockerfiles/stages/3-opam-deps:2]: what's the workflow when I'm a dev who wants to change our opam deps? do I rebuild this and push it?

File: dockerfiles/stages/4-deb-builder [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    src/app/cli/src/mina_mainnet_signatures.exe \
+    src/app/archive/archive.exe \
+    src/app/archive/archive_testnet_signatures.exe \
+    src/app/archive/archive_mainnet_signatures.exe \
+    src/app/rosetta/rosetta.exe \
+    src/app/rosetta/rosetta_mainnet_signatures.exe \
+    src/app/rosetta/rosetta_testnet_signatures.exe \
+    src/app/rosetta/test-agent/agent.exe \
+    src/app/rosetta/ocaml-signer/signer.exe
+
```
Comment (bkase) [inline dockerfiles/stages/4-deb-builder:49]: you have two 4- files, is this intentional?

File: buildkite/src/Command/TestExecutive.dhall (hunk 2 above)
Comment (bkase) [reply buildkite/src/Command/TestExecutive.dhall:7]: nevermind -- I see 👍

File: dockerfiles/stages/3-opam-deps (hunk 4 above)
Comment (bkase) [reply dockerfiles/stages/3-opam-deps:2]:
```
Remove opam.tar.gz caching and instead include all build-time dependencies in MinaToolchain
dockerfiles/stages/2-toolchain
dockerfiles/stages/3-opam-deps
buildkite/src/Command/RunInToolchain.dhall
buildkite/src/Command/OpamInit.dhall (deleted)
Still some work remaining to automatically rebuild this when there are new dependencies. CI will build it, but you still have to manually update buildkite/src/Constants/ContainerImages.dhall
```

^^ cool CI does it for you 💯

File: buildkite/scripts/generate-diff.sh (hunk 1 above)
Comment (nholland94) [reply buildkite/scripts/generate-diff.sh:3]: Shouldn't this default to `compatible` instead of `master`?

File: dockerfiles/stages/1-build-deps [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+ARG RUST_VERSION=1.52.1
+
+# Rocksdb commit tag/branch to clone
+ARG ROCKSDB_VERSION=v5.17.2
+
+# OS package dependencies
+# First add support for https and pkg-config for apt, then install everything else
+# TODO: make sure this is the minimum compile-time deps
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update \
```
Comment (nholland94) [inline dockerfiles/stages/1-build-deps:28]: Shouldn't we remove the apt cache after this step?

File: dockerfiles/stages/3-opam-deps [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+WORKDIR $HOME/$MINA_DIR
+
+ENV OPAMYES 1
+
+# TODO: handle this opam work without cloning the full repository (directly pull src/opam.export)
+RUN opam switch import src/opam.export
+
+# TODO: Get pins from a script that can be easily updated if more pins are needed
+# Would be really nice to pull this script, the git submodules, and src/opam.export exclusively in this stage
+RUN eval $(opam config env) \
```
Comment (nholland94) [inline dockerfiles/stages/3-opam-deps:40]: Can we use `./scripts/pin-external-packages.sh`?

**Review by nholland94 — APPROVED** [review]
Looks great overall, just a few small comments.

File: dockerfiles/stages/1-build-deps (hunk 6 above)
Comment (nholland94) [reply dockerfiles/stages/1-build-deps:28]: Same with other stages that do apt-get installs.

File: dockerfiles/stages/4-deb-builder (hunk 5 above)
Comment (lk86) [reply dockerfiles/stages/4-deb-builder:49]: yes, this one fits in as a replacement for 4-builder at the same point in the set of stages (both 4- stages start from the 3- container and they both build similar artifacts, just not in the same way).

File: dockerfiles/stages/1-build-deps (hunk 6 above)
Comment (lk86) [reply dockerfiles/stages/1-build-deps:28]: No actually, in `slim` debian images apt is already configured to skip creating/maintaining that cache in the first place so we don't need to delete it afterwards.

**Review by bkase — APPROVED** [review]
Looks great!

**Pull request merged**